### PR TITLE
do not insert to_dtype for memory copy only buffers

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97147

 Remove redundant to_dtype like 
`load_bf16 + to_fp32 + to_bf16 + store_bf16` => `load_bf16 + store_bf16`

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1148,7 +1153,11 @@ def load(self, name: str, index: sympy.Expr):
             )
             line = f"at::vec::Vectorized<float>::loadu({self.var_vec_buf_map[var]})"
         elif V.graph.get_dtype(name) in [torch.bfloat16]:
-            line = f"load_bf16_as_float({var_expr})"
+            if opt_ctx.is_load_bf16_as_fp32:
+                line = f"load_bf16_as_float({var_expr})"
+            else:
+                assert opt_ctx.is_directly_load_bf16
+                line = f"at::vec::Vectorized<bfloat16>::loadu({var_expr}, {self.tiling_factor})"
```
Comment (EikanWang) [inline torch/_inductor/codegen/cpp.py:1160]: If the `dtype` is `bfloat16`, will the `tiling_factor` be 32 for AVX512 and 16 for AVX2?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -1651,11 +1666,14 @@ def is_load_only_block(self, sub_graph: torch.fx.Graph):
                     break
 
                 # Create and record the context
-                opt_ctx = OptimizationContext()
+                if OptimizationContext.key in _node.meta:
+                    opt_ctx = _node.meta[OptimizationContext.key]
```
Comment (EikanWang) [inline torch/_inductor/codegen/cpp.py:1670]: Does that mean the `store` node could be executed multiple times?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -1994,10 +2012,39 @@ def __init__(self, kernel_group):
...
+                    if store_node.target == "load"
+                    else store_node.args[-1]
+                )
+                if only_used_by_store(store_node) and store_dtype == load_dtype:
+                    opt_ctx = OptimizationContext()
+                    opt_ctx.is_directly_store_bf16 = True
+                    node.meta[OptimizationContext.key] = opt_ctx
+                    return True
+                else:
+                    return False
```
Comment (EikanWang) [inline torch/_inductor/codegen/cpp.py:2040]: Could we combine the two functions as a single one? We just need to support the loop body is memcpy.

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -1994,10 +2012,39 @@ def __init__(self, kernel_group):
...
         def add_to_dtype(sub_graph: torch.fx.Graph):
+            def only_used_by_store(node: torch.fx.Node):
+                if len(node.users) > 1:
+                    return False
+                for user in node.users:
+                    _only_used_by_store = user.target == "store"
+                    opt_ctx = OptimizationContext()
+                    opt_ctx.is_directly_load_bf16 = _only_used_by_store
+                    node.meta[OptimizationContext.key] = opt_ctx
+                    return _only_used_by_store
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2023]: Is it fine regardless of how many users as long as the users are stores and the store dtype is bfloat16?

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:2040]: Also, what's the difference between "is_directly_load_bf16" and "is_directly_store_bf16"? Do they serve the same purpose?

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:1160]: No, the `tiling_factor` will be 16/8 for AVX512/AVX2. This PR just to remove the `to_dtype`, still only load half vec-size for `BF16`.
```
tiling_factor = self.picked_vec_isa.nelements(dtype=torch.float)
```

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:1670]: I think the `store` node may have `opt_ctx` in `legalize_bf16`. So need check whether already have one here.

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py]: Yes, thanks for advice.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py]: Yes, the two functions and two flags can be combined as a single one

File: torch/_inductor/codegen/cpp.py [hunk 5]
```diff
@@ -1994,10 +2011,35 @@ def __init__(self, kernel_group):
...
+                    )
+                    if bf16_mem_copy:
+                        opt_ctx = OptimizationContext()
+                        opt_ctx.is_bf16_mem_copy = bf16_mem_copy
+                        node.meta[OptimizationContext.key] = opt_ctx
+                    return bf16_mem_copy
+                elif node.target is ["store"]:
+                    stored_node = node.args[3]
+                    bf16_mem_copy = is_bf16_mem_copy(stored_node)
+                else:
```
Comment (EikanWang) [inline torch/_inductor/codegen/cpp.py:2024]: The target of a node could be `get_index` and the semantic is still `memcpy`. Please check it.

Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:2024]: I think we do not need care about `get_index`.
 - we only invoke `is_bf16_mem_copy` to check whether need to insert `to_dtype` when `node.target in ("load", "store", "constant")`.
 - Maybe we should assert it in `is_bf16_mem_copy`?

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:1159]: We may consider to improve this further as the next step. In general, we don't have to care about what data type is when we do memory copy. Maybe we can do some pattern matching on the kernel fx body to figure out if it is a memory copy and just apply a common memory copy intrinsics.

File: torch/_inductor/metrics.py [hunk 6]
```diff
@@ -7,16 +7,21 @@
 # counters for tracking fusions
 ir_nodes_pre_fusion = 0
 
+# counters for tracking to_dtype inserted
+to_dtype_count = 0
```
Comment (jgong5) [inline torch/_inductor/metrics.py:11]: If it is cpp specific, add "cpp" prefix?

Comment (zhuhaozhe) [reply torch/_inductor/metrics.py]: Done

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:1159]: Agree, this should be done in a larger scope (FX body). But we also need to consider not to loss fusion opportunity. Will be a next step.

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (EikanWang) [reply torch/_inductor/codegen/cpp.py:1159]: > We may consider to improve this further as the next step. In general, we don't have to care about what data type is when we do memory copy. Maybe we can do some pattern matching on the kernel fx body to figure out if it is a memory copy and just apply a common memory copy intrinsics.

Regarding this optimization, we do not need to use vectorization but invoke C/C++ `memcopy` directly.

**Review by EikanWang — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Comment by jgong5** [comment]
@zhuhaozhe BTW, transposed 2d copy only supports float currently. Will you add bfloat16 support later?

**Comment by zhuhaozhe** [comment]
> @zhuhaozhe BTW, transposed 2d copy only supports float currently. Will you add bfloat16 support later?

Will take a look later.

**Pull request closed without merging**